include/engine/api/route_api.hpp

```diff
@@ -215,43 +215,71 @@ class RouteAPI : public BaseAPI
+                            durations.values.push_back(step.duration);
+                        });
+                    annotation.values["duration"] = std::move(durations);
+                }
+                if (parameters.annotations_type & RouteParameters::AnnotationsType::Distance)
+                {
+                    util::json::Array distances;
+                    distances.values.reserve(leg_geometry.annotations.size());
+                    std::for_each(
+                        leg_geometry.annotations.begin(),
+                        leg_geometry.annotations.end(),
+                        [this, &distances](const guidance::LegGeometry::Annotation &step) {
+                            distances.values.push_back(step.distance);
+                        });
+                    annotation.values["distance"] = std::move(distances);
+                }
+                if (parameters.annotations_type & RouteParameters::AnnotationsType::Weight)
+                {
+                    util::json::Array weights;
+                    weights.values.reserve(leg_geometry.annotations.size());
+                    std::for_each(leg_geometry.annotations.begin(),
+                                  leg_geometry.annotations.end(),
+                                  [this, &weights](const guidance::LegGeometry::Annotation &step) {
+                                      weights.values.push_back(step.weight);
+                                  });
+                    annotation.values["weight"] = std::move(weights);
+                }
+                if (parameters.annotations_type & RouteParameters::AnnotationsType::Datasources)
+                {
+                    util::json::Array datasources;
+                    datasources.values.reserve(leg_geometry.annotations.size());
+                    std::for_each(
+                        leg_geometry.annotations.begin(),
+                        leg_geometry.annotations.end(),
+                        [this, &datasources](const guidance::LegGeometry::Annotation &step) {
+                            datasources.values.push_back(step.datasource);
+                        });
+                    annotation.values["datasources"] = std::move(datasources);
+                }
+                if (parameters.annotations_type & RouteParameters::AnnotationsType::Nodes)
+                {
+                    util::json::Array nodes;
+                    nodes.values.reserve(leg_geometry.osm_node_ids.size());
+                    std::for_each(leg_geometry.osm_node_ids.begin(),
+                                  leg_geometry.osm_node_ids.end(),
+                                  [this, &nodes](const OSMNodeID &node_id) {
+                                      nodes.values.push_back(static_cast<std::uint64_t>(node_id));
+                                  });
+                    annotation.values["nodes"] = std::move(nodes);
+                }
```

Comment:
Hm the ~100 lines above are all doing the same:
- check if annotation x was requested
- create `json::Array` for many x
- extract `step.x` from `RouteStep` and add into array
- add array to annotations response as `"x" : [...]`

I think you could abstract over this and make it extensible for the future at the same time.

What about abstracting this behind a function which takes
- a `AnnotationsType`
- a lambda taking a `RouteStep` and returning annotations items out of it (e.g. `step.duration` for `AnnotationsType::Duration`
- a string describing the annotations type for the json object's key as in `"key": [...]` above

Example callsite:

```c++
makeAnnotation(AnnotationsType::Duration, "duration", [](auto& step){ return step.duration });
```

which then does all of the steps above behind the scenes and wires the parts together.